File: haskell/binary-search-tree/BST.hs
module BST (bstLeft, bstRight, bstValue, singleton, insert, fromList, toList) where

import Data.List (foldl')

data Node a = Node { bstLeft :: Maybe (Node a), bstRight :: Maybe (Node a), bstValue :: a }

singleton :: a -> Node a
singleton value = Node { bstLeft = Nothing, bstRight = Nothing, bstValue = value }

insert :: (Ord a) => a -> Node a -> Node a
insert newValue tree 
    | newValue <= bstValue tree = tree { bstLeft = insertLoop newValue (bstLeft tree) }
    | otherwise = tree { bstRight = insertLoop newValue (bstRight tree) }

insertLoop :: (Ord a) => a -> Maybe (Node a) -> Maybe (Node a)
insertLoop newValue node = case node of
    Just tree -> Just (insert newValue tree)
    Nothing -> Just (singleton newValue)

fromList :: (Ord a) => [a] -> Node a
fromList [] = error "Cannot create tree from empty list"
fromList (x:xs) = foldl' (flip insert) (singleton x) xs

toList :: Node a -> [a]
toList = toListLoop . Just

toListLoop :: Maybe (Node a) -> [a]
toListLoop node = case node of
    Just (tree) -> toListLoop (bstLeft tree) ++ [bstValue tree] ++ toListLoop (bstRight tree)
    Nothing -> []

File: v1/scala/binary-search-tree/src/main/scala/Bst.scala
case class Bst[T: Ordering](value: T, left: Option[Bst[T]], right: Option[Bst[T]]) {
  def insert(newValue: T): Bst[T] =
    if (Ordering[T].lteq(newValue, value))
      new Bst(value, insert(newValue, left), right)
    else
      new Bst(value, left, insert(newValue, right))

  private def insert(newValue: T, tree: Option[Bst[T]]): Option[Bst[T]] = tree match {
    case Some(branch) => Some(branch.insert(newValue))
    case None => Some(Bst(newValue))
  }
}

object Bst {
  def apply[T: Ordering](value: T): Bst[T] = new Bst(value, None, None)

  def fromList[T: Ordering](values: List[T]) = values match {
    case x :: xs => xs.foldLeft(Bst(x))((tree, value) => tree.insert(value))
    case Nil => throw new IllegalArgumentException("Cannot create tree from empty list")
  }

  def toList[T: Ordering](tree: Bst[T]): List[T] = toList(Some(tree))

  private def toList[T: Ordering](tree: Option[Bst[T]]): List[T] = tree match {
    case Some(x) => toList(x.left) ++ List(x.value) ++ toList(x.right)
    case None => Nil
  }
}